stripe/src/main/java/com/stripe/android/view/PaymentMethodsActivity.java

```diff
@@ -0,0 +1,363 @@
+        if (getSupportActionBar() != null) {
+            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+        }
+
+        if (localCustomer != null) {
+            mCustomer = localCustomer;
+            initWithCustomer();
+            return;
+        }
+
+        boolean waitForProxy = getIntent().getBooleanExtra(EXTRA_PROXY_DELAY, false);
+        if (waitForProxy) {
+            return;
+        }
+        initializeData();
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        if (requestCode == REQUEST_CODE_ADD_CARD && resultCode == RESULT_OK) {
+            setCommunicatingProgress(true);
+            CustomerSession.CustomerRetrievalListener listener =
+                    new CustomerSession.CustomerRetrievalListener() {
+                        @Override
+                        public void onCustomerRetrieved(@NonNull Customer customer) {
+                            addSelectionIfNecessaryAndUpdateCustomer(customer);
+                        }
+
+                        @Override
+                        public void onError(int errorCode, @Nullable String errorMessage) {
+                            String displayedError = errorMessage == null ? "" : errorMessage;
+                            showError(displayedError, mCustomerSessionProxy == null);
+                            setCommunicatingProgress(false);
+                        }
+                    };
+            if (mCustomerSessionProxy == null) {
+                CustomerSession.getInstance().updateCurrentCustomer(listener);
+            } else {
+                mCustomerSessionProxy.updateCurrentCustomer(listener);
+            }
+        }
+    }
+
+    @Override
+    public boolean onPrepareOptionsMenu(Menu menu) {
+        MenuItem saveItem = menu.findItem(R.id.action_save);
+        Drawable tintedIcon = ViewUtils.getTintedIcon(
+                this,
+                R.drawable.ic_checkmark,
+                android.R.color.primary_text_dark);
+        saveItem.setIcon(tintedIcon);
+        return super.onPrepareOptionsMenu(menu);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        // Inflate the menu; this adds items to the action bar if it is present.
+        getMenuInflater().inflate(R.menu.add_source_menu, menu);
+        menu.findItem(R.id.action_save).setEnabled(!mCommunicating);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == R.id.action_save) {
+            setSelectionAndFinish();
+            return true;
+        } else {
+            boolean handled = super.onOptionsItemSelected(item);
+            if (!handled) {
+                onBackPressed();
+            }
+            return handled;
+        }
+    }
+
+    void addSelectionIfNecessaryAndUpdateCustomer(@NonNull Customer customer) {
```

Comment:
nit: updateCustomerAndAddDefaultSourceIfNecessary for clarity

I totally get that the logic here is complex and I'm sure you've spent cycles making this more clear. Haha, but I still had some trouble parsing it so any additional efforts to improve clarity would be appreciated 

I personally bias for fewer comments but I would be into something along the lines with "this function updates the customer with a source. if that source is the only source, we set it as the default source"